### ISSUE TITLE
About the CasMVSNet

### ISSUE THREAD

**Issue by whubaichuan** [kwea123/CasMVSNet_pl#2]
Hi, recently I read the paper "Deep Stereo using Adaptive Thin Volume Representation
with Uncertainty Awareness" published on 27 Nov 2019, see the [paper](https://arxiv.org/pdf/1911.12012v1.pdf). I think CasMVSNet is so similar to that paper.

**Comment by kwea123** [comment]
In my opinion the main difference is that CasMVSNet use fixed depth range for each pixel and that paper use pixel-wise depth range depending on the confidence. I've read [this](https://arxiv.org/pdf/1909.03751.pdf) which seems to have a similar concept; it seems using pixel-wise things is a good idea.
I feel that the current CasMVSNet's precision is still limited by the `depth_interval` parameter, imagine in the finest stage the model predicts `0.5` proba for `d` and `0.5` proba for `d+1`, the model can only be precise up to `depth_interval/2`. Although thanks to the fact that the model uses small GPU memory, so we can decrease the `depth_interval` and use more planes/increase `interval_ratio` for coarser stages, the global precision is still limited by this value. On the other hand, if we let the model learn the uncertainty and adaptive depth range, the `depth_interval` of the finest stage can now be very small, if the model is confident enough.

That is my impression. I might give it a try in the future, for now I'm trying to adapt the model for other datasets: ETH3d/Tanks and Temples/BlendedMVS inference or training.

**Comment by kwea123** [comment]
I tried to implement the pixel-wise depth interval using std in "Deep Stereo using Adaptive Thin Volume Representation with Uncertainty Awareness", but the result is worse than the current result.
I tried both 
1. Simply replace depth interval with std value without any training
2. Replace depth interval with std value with training
but both are bad.

After inspecting the depth confidence, I found that there exists many peaks in fine level predictions meaning the network is very uncertain, in contrast to Fig.3 in the paper which seems to have only one peak.
With the findings of another [paper](https://github.com/JiayuYANG/CVP-MVSNet), they state that the pixel difference of different depth plane must be at least 0.5pix to make the pixel feature distinguishable (=more certain of which corresponding pixel is correct). However, if the depth interval can be reduced to 0.836mm like stated above, I made some statistics and it would mean **most of** the pixel differences can be as small as 0.2 pix (size 640x512 is fixed)! With my current knowledge it is nearly impossible to distinguish two pixels with only that small difference unless the feature extraction is super powerful. In their paper however it looks like it's almost the same as what I have here, so I have no idea how distinguishing 0.2pix is possible.
I will need to look at their final implementation to conclude, but currently I cannot reproduce their result and am suspicious about the actual effect.

**Comment by whubaichuan** [comment]
@kwea123 The two projects you mentioned are all opensource. [CVP-MVSNet](https://github.com/JiayuYANG/CVP-MVSNet) and [UCSNet](https://github.com/touristCheng/UCSNet). Can you find the concrete implementation about how to distinguishing 0.2pix in the code?  Do the two projects reach the results as their paper said?

**Comment by kwea123** [comment]
UCSNet not open sourced yet, so I cannot tell whether my implementation is correct or not. The 0.836mm=0.2pix is based on my calculation on one image only so maybe not that correct.
I haven't tried CVP-MVSNet yet but their result seems more reasonable.

**Comment by kwea123** [comment]
Another interesting finding: instead of the pixel-wise depth interval, I use the current same depth interval for all pixels, but try different depth interval for **the finest level** while maintaining the depth interval for other levels, in order to find the optimal depth interval. More precisely, I set the depth intervals to be *around* `X 5.5 10.5` for the three levels, where I want to find the optimal `X`. The other hyperparameters are set the same.
The result on the val set of DTU is as follows:
|    X     | abs_err (mm) | acc_1mm (%) | acc_2mm (%)   | acc_4mm (%)    |
| :---:    | :---:      |  :---:          | :---:            | :---:             | 
| 2.65   |  4.41   | 70.49     |  83.60     | 90.41   |
| 2.00   |  4.23   | 70.69     |  83.59     | 90.37   |
| 1.50   |  **4.20**   | **71.02**     |  **83.75**     | **90.52**   |
| 1.25   |  4.44   | 70.68     |  83.27     | 89.97   |
| 1.00   |  4.76   | 69.67     |  82.41     | 89.33   |

When we decrease the depth interval, intuitively we expect that we can have higher accuracy, however, as can be seen in the table, it turns out **it is not the case**, decreasing the depth interval too much degrades the performance in contrast to our intuition. The optimal `X` turns out to be somewhere around 1.25 and 1.5 for this `X 5.5 10.5` interval and `8 32 48` n_depths settings.

In the implementation, finer level results have marginal impact on coarser level results due to the `.detach()` of depth map (but still exists in image FPN), so in my opinion the different results for different `X`s simply means if we set `X` too small, it will cause regression error on the finest level only, while the coarser level results are good (this needs careful evaluation on the coarser levels results which I haven't done).

To sum up, based on the above table and my failure on reproducing the results on [Deep Stereo using Adaptive Thin Volume Representation with Uncertainty Awareness](https://github.com/touristCheng/UCSNet), I am more believing in the claim of CVP-MVSNet which says the depth interval should not be too small (at least 0.5pix difference, although I still don't know how much mm is that). I will do more inspection on the coarse level results to confirm this finding and wait for the open source of UCSNet before coming back to this.

**Comment by flamehaze1115** [comment]
Hi. When I use pretrained model provided by CasMvsNet to test on tanks and temple, I find that there will be too many background pixels on final point cloud. Do you meet such problem? And the background pixels will influence the f score or not?

**Comment by cvgogogo** [comment]
I also want to know those tricks @whubaichuan 😃

**Comment by whubaichuan** [comment]
@cvgogogo The photo_threshold plays a vital role. I set the photo_threshold to 0.6 and get the better f-score. The different depth-interval will lead to different accuracy but I can not get a better f-score. So the suggestion is trying to change the photo_threshold.
